### DISCUSSION TITLE
getServerSideProps error: "Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value all together."

### DISCUSSION THREAD

**Discussion by jgpacker** [vercel/next.js#11209] [Help]
The `undefined` type is convenient to use in Typescript, specially because of the optional-attributes feature.

However, in the code below:
```typescript
import React from 'react';
import { NextPage, GetServerSideProps } from "next";

const getA = (): number => 3;

const Test: NextPage<{a?: number}> = ({ a }: {a?: number}): JSX.Element => {
  return (<>
    <p>{a}</p>
  </>);
}

export const getServerSideProps: GetServerSideProps<{a?: number}> = async ()  => {
  const a = getA();
  return { props: { a: a > 5? a: undefined} };
}

export default Test;
```
We get this error :
```
Error serializing `.a` returned from `getServerSideProps` in "/test". Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value all together.

Error: Error serializing `.a` returned from `getServerSideProps` in "/test".
Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value all together.
```
I can imagine this kind of validation was added for a good reason: to avoid lost hours debugging why a prop that is not JSON-serializable was not received by the component.

However, it would be interesting to allow `undefined`, and omit it from the resulting JSON.

**Comment by bradennapier** [comment]
I agree - in addition this becomes a problem when we use query strings which may be undefined then inject them back into props as optional values as well as other cases where we may have undefined as a result.

At the very least we should be able to allow `undefined` to be allowed in props.  It is a bit odd to iterate these objects just to remove undefined values and/or to have to resort to object mutation here to make a perfectly clean object?

I also understand the general reasoning for this... but it will get pretty verbose in many cases.  it appears using an object for this does fix it but ... 


```typescript
export const getServerSideProps: GetServerSideProps = async ({ query }) => {
  return {
    props: {
      alertMessage: query.alertMessage || null,
      alertMessageType: query.alertMessageType || null,
      one: 1,
    },
  };
};
```

**Reply by kaoutharsalah** [reply]
solution with JS please

**Reply by zy-jtp** [reply]
remove `: GetServerSideProps`. Also i suggest you study a bit of TS, is not hard, converting from TS is literally deleting code, 100% of the time. Also [this tool](https://transform.tools/typescript-to-javascript)

**Comment by lkbr** [comment]
I understand the examples given might be simplified versions of your use cases, but perhaps you might be overthinking this? If something is undefined, don't include it in GetServerSideProps props object, then it will be undefined in the Page props.

**Comment by berrigan** [comment]
The problem is that omitting `undefined` values you already have from upstream/data can be difficult. It's essentially asking the `getServerSideProps` code to do sanitization for `undefined`, when the ensuing ~ `JSON.stringify()` to serialize the props would also do the removal anyway.

Or, worse, re-writing the upstream data to not use `undefined`, which may also be impossible if it's a 3rd party. In my case, it's data from our server, so doable - but seems quite unnecessary.

Maybe a console warning would be better instead of blocking development.

In one case I've done recently I also intentionally have a value which is `true | undefined`, knowing that the `undefined` case will simply be omitted by the JSON, allowing for the props to be smaller than a straight `boolean`.

**Comment by jgpacker** [comment]
One issue is that `undefined` and `null` only have subtle differences, and it can be interesting to only use `undefined` in TypeScript for consistency's sake and because TypeScript's optional parameters/attributes makes it convenient to use `undefined` instead of `null`.

**Reply by kendallstrautman** [reply]
+1 here, clean typescript interfaces often use the optional param which defaults as `undefined`. Adjusting all these optional properties for a verbose (fairly unnecessary ) alternative. 

```
interface Data {
    title: string
    description?: string
    rank?: number
}

// versus

interface Data {
    title: string
    description: string | null
    rank: number | null
}
``` 

And imagine this for large complex, deeply nested interfaces that often propagate page level data fetching. 

Additionally, using `null` instead of `undefined` upsets the pattern of setting default props for optional params.

**Reply by tonytamps** [reply]
Your two examples aren't equivalent @kendallstrautman 

`description?: string` says the key may not be present but if it is present then it will contain a `string` value.
`description: string | null` says the key will always be present and will either be a `string` or `null`. 

I use both where appropriate. These shouldn't be treated the same by Next nor JS.

**Reply by kendallstrautman** [reply]
@tonytamps  Yes I understand they aren't the same, that is the point. This latter example shows the alternative interface needed to align with the issue in this thread — that Next.js data fetching doesn't allow `undefined` as props. The first example shows the preferred interface.

**Comment by xcrzx** [comment]
In my case, it's a deeply nested data from a 3rd party service. I use the following workaround to bypass this issue:
```typescript
export const getServerSideProps: GetServerSideProps<PageProps> = async (ctx) => {
  const props = await getPropsFromAService()

  return { props: JSON.parse(JSON.stringify(props)) }
}
```
However, it does look really awkward.

**Reply by jgpacker** [reply]
Another possible workaround: 
```js
// Hack needed to avoid JSON-Serialization validation error from Next.js https://github.com/zeit/next.js/discussions/11209
// >>> Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value all together.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const deleteUndefined = (obj: Record<string, any> | undefined): void => {
  if (obj) {
    Object.keys(obj).forEach((key: string) => {
      if (obj[key] && typeof obj[key] === 'object') {
        deleteUndefined(obj[key]);
      } else if (typeof obj[key] === 'undefined') {
        delete obj[key]; // eslint-disable-line no-param-reassign
      }
    });
  }
};
```

**Reply by jerson100** [reply]
👍

**Reply by bombillazo** [reply]
@xcrzx That did it for us, looks super hacky but it works!

**Reply by Christopher-Hayes** [reply]
I like the `deleteUndefined()` method since it should be many times faster, particularly if you're working with large content objects from a CMS.

**Reply by tranduclinh2067** [reply]
thank you my love

**Comment by ivawzh** [comment]
A better alternative doing it would be NextJs (optionally) allows `undefined`. And then use mapped types like `InferGetStaticPropsType` to sanitize and pick non-undefined fields only. 

I can help open PR if Nextjs agrees with the approach

**Comment by ale-grosselle** [comment]
Hi folks!
In terms of performance which is the best workaround to:
`JSON.parse(JSON.stringify(props))`
or what propose @jgpacker here: https://github.com/vercel/next.js/discussions/11209#discussioncomment-38480


Many thanks for the support!

**Reply by jgpacker** [reply]
It seems this JSON validation only happens in development mode, so performance would not be an issue in production mode:
```js
// inside getServerSideProps
  if (process.env.NODE_ENV !== 'production') {
    deleteUndefined(myProps);
  }
```
However note that it's recommended to avoid verifying `NODE_ENV` to reduce inconsistencies* between development and production modes.
\* a possible inconsistency: `JSON.stringify([1, undefined, 2])` results in `"[1,null,2]"`

**Reply by Chesare22** [reply]
I don't know what is faster, but the performance benefit would be too small to notice. It is better to ask "what will be easier to troubleshoot if something went wrong?". I prefer `JSON.parse(JSON.stringify(props))` for the simplicity.

**Comment by ChristianIvicevic** [comment]
I am solving this issue using the spread operator and conditionally spread my possibly undefined values only when they do in fact exist. This is not really pretty, but IMHO much cleaner than using `JSON.parse(JSON.stringify(...))` or the proposed `deleteUndefined ` method.

```typescript
return {
  props: {
    currentPost: allPosts[currentPostId],
    ...(previousPost === undefined
      ? {}
      : {
          previousLink: {
            href: previousPost.slug,
            title: previousPost.title,
          },
        }),
    ...(nextPost === undefined
      ? {}
      : {
          nextLink: {
            href: nextPost.slug,
            title: nextPost.title,
          },
        }),
  },
};
```

**Comment by damacisaac** [comment]
I agree with @ivawzh that none of the workarounds here are sufficient. They push a lot of complexity to `getServerSideProps` and it's difficult to protect yourself against this using types. The chances of accidentally introducing an undefined value somewhere, which is an extremely common occurance, is too high.

**Comment by lafpina** [comment]
I am running into this problem in the middle of a refactoring from React to Next. It's been a time consuming trying to figure out the problem to find out a solution. I got stuck on that but now I am getting some light from these posts. Thanks all for that!

**Comment by tkrotoff** [comment]
FYI A pull request has been proposed https://github.com/vercel/next.js/pull/17209 on Sep 18, 2020 to allow undefined to be serializable

**Reply by fauskanger** [reply]
It seems the PR is closed - does anyone know what the issue with it is? Seems it didn't pass the tests? A very welcomed upgrade if it works, though

**Comment by kelly-tock** [comment]
just adding to this, nextjs should also consider apps with a large redux footprint and typescript, with the `undefined` case being a pretty common convention. 

initial state:
```
value: undefined
```

and typed as 
```
value?: string;
```

super common in typescript redux apps.

**Reply by kelly-tock** [reply]
Another area where this happens as well is using with next-redux-wrapper, which is a bit different in how it works.

**Reply by kelly-tock** [reply]
I know I may be stating the obvious here, but I ended up deciding to go with `getInitialProps`, and it doesn't seem to have this issue at all?  The reason I went with that was that I have some initial app data, config, that I need to fetch once on the server side, and then while navigating I don't want to fetch it again. This allows me to know client side if the config is already present to just not fetch it again. 

I guess getInitialProps doesn't go through the same check?

**Comment by themagickoala** [comment]
I'm going to add to this - I have GraphQL results coming back from an API, which means that my response types are necessarily sparse. Having to sanitise this to null for every use case is a massive overhead that I don't want to have to maintain. At the very least a "dangerous" mode that can be opted in to that filters out undefined values would save a huge amount of error-prone additional work.

**Comment by kachkaev** [comment]
If you use TypeScript and `undefined` is only a problem on the top-level, this workaround might work:

```ts
const removeUndefinedForNextJsSerializing = <T,>(props: T): T =>
  Object.fromEntries(
    Object.entries(props).filter(([, value]) => value !== undefined),
  ) as T;

export const getStaticProps: GetStaticProps<MyPageProps> = async () => {
  return {
    props: removeUndefinedForNextJsSerializing({
      myProp1: await getSomethingThatCanBeUndefined(),
      myProp2: await getSomethingThatCanBeUndefined(),
    }),
  };
};
```

I wonder if Next.js could apply this fix by itself for the direct descendants of `props` 🤔

**Comment by mattcasey** [comment]
Been loving next.js so far and how it finds a good balance of opinionated development. However, in this case, I feel like the library is being a bit too picky. I was able to monkey patch the library globally by adding the following in `next.config.js`. Beware if you use this, it's possible it could break in a future update.
```
/**
 * Remove undefined values so Next.js doesn't complain during serialization. Verified as of v11.1.2.
 */
const removeUndefined = (obj) => {
  let newObj = {};
  Object.keys(obj).forEach((key) => {
    if (obj[key] === Object(obj[key])) newObj[key] = removeUndefined(obj[key]);
    else if (obj[key] !== undefined) newObj[key] = obj[key];
  });
  return newObj;
};
const next = require('next/dist/lib/is-serializable-props');
const isSerializableProps = next.isSerializableProps;
next.isSerializableProps = (page, method, input) => isSerializableProps(page, method, removeUndefined(input));
```

**Comment by kasparkallas** [comment]
Worth mentioning is that Redux, the incredibly popular state container which is often used together with Next.js, also has a guideline/constraint for data to be serializable... and it happily allows `undefined`. Considering the popularity of these projects in the React ecosystem, I think it would be beneficial for the projects to treat serializability the same.

If anyone is running into the serialization issue with Redux and using `next-redux-wrapper` then the solution is following:
```ts
export const wrapper = createWrapper<AppStore>(makeStore, {
    debug: true,
    serializeState: (state) => JSON.stringify(state),
    deserializeState: (state) => JSON.parse(state),
});
```
source: https://github.com/kirill-konshin/next-redux-wrapper/issues/276#issuecomment-891342315

**Comment by hugejile** [comment]
A stupid solution is here

```typescript
export const getServerSideProps = withSessionServerSide(async (ctx: GetServerSidePropsContext) => {
    return {
        props: { auth: ctx.auth },
    }
})

export function withSessionServerSide<P extends { [key: string]: unknown } = { [key: string]: unknown },>(
    handler: (context: GetServerSidePropsContext,) => GetServerSidePropsResult<P> | Promise<GetServerSidePropsResult<P>>,
) {
    return async function nextGetServerSidePropsHandlerWrappedWithIronSession(
        ctx: GetServerSidePropsContext,
    ) {
        Object.defineProperty(ctx, 'auth', {
            value: { accessToken, accessTokenExpiredIn },
        });

        if (process.env.NODE_ENV !== 'production') {
            //Check property is undefined only happens in development mode
            const r = await handler(ctx);
            return JSON.parse(JSON.stringify(r))
        }
        else {
            return handler(ctx)
        }
    };
}

**Comment by albertstill** [comment]
Some issues I have with classing `undefined` as unserializable.

- `JSON.parse(JSON.stringify({foo: undefined})).foo === undefined`, the JSON string not containing the field means JS will return undefined on read, so in a lot of cases the code behaves as expected in the front end.
- Suggesting using `null` as an alternative to `undefined` means React's [`defaultProps`](https://reactjs.org/docs/typechecking-with-proptypes.html#default-prop-values) will not work, default props are only used when the prop being passed in is `undefined`, therefore `null` would go through to the React component and potentially cause FE bugs. 
- Omitting fields from JS object literals in src is cumbersome as you have imperatively build the JS object with statements like `if (foo) obj.foo = foo;` as appose to an object literal and an inline ternary `foo: foo ? foo : undefined, // undefined to trigger defaultProps`.
- Any FE JS that depends on an object key being present albeit `undefined` or `null`, e.g use of `Object.keys` or `hasOwnProperty`, will need to use a `null` value on the server. The check in this case is useful but I think these cases are not common enough to outweigh the above issues.

For our server I have added `deleteUndefined` function that runs in dev only so I can add Next to our exisiting server's prop generation but I think it would be nice if we could opt out of this check for `undefined`'s via `next.config.js`.

**Comment by MichaelAllenWarner** [comment]
At the very least, it would be nice if `getStaticProps` and `getServerSideProps` would give TypeScript warnings if you try to use a possibly `undefined` value in the returned `props`. Is that feasible? It would help users catch this mistake immediately in development instead of running into a build-error down the line.

**Comment by vsDizzy** [comment]
Correct me if I'm wrong. But the only reason of this weird check is to slow down the development with implementing a workaround.
- The message is wrong. If you try to serialize undefined in the console you will have no error. 🧐
- The reason of this check is unclear. And there is no need to block Javascript powers. 🧙‍♂️
- Suggestion to use null is weird because the `img` component expect `src` to be `string | undefined`. Your `Img` control is a bad and broken, so I'm using `img` wherever I can. 😎

Please remove this check (and your bad and broken `Img` control).

**Comment by fbartho** [comment]
Is this a feature flag we can opt out of? It's frustrating that I can't pull in the libraries I want just because they happen to work with `undefined` fields.

**Reply by fbartho** [reply]
For my codebase, I've added a postinstall script that fixes this:

```typescript
const rootDir = path.resolve(__dirname, "..");
const isSerializablePropsPath = path.join(
	rootDir,
	"node_modules",
	"next",
	"dist",
	"lib",
	"is-serializable-props.js",
);
const patchRegexp =
	/\s(throw new SerializableError\(page, method, path, '`undefined` cannot be serialized as JSON)/;
function main() {
	let content = fs.readFileSync(isSerializablePropsPath, "utf8").toString();

	if (patchRegexp.test(content)) {
		content = content.replace(patchRegexp, "return true;//$1");
		fs.writeFileSync(isSerializablePropsPath, content);
		console.debug("Patched next/lib::is-serializable-props to allow 'undefined' values.");
	} else {
		console.debug("Already patched (next/lib::is-serializable-props to allow 'undefined' values).");
	}
}
```

**Reply by solomonhawk** [reply]
@fbartho If you need to resort to editing dependency code in `node_modules` you might be interested in [patch-package](https://github.com/ds300/patch-package) as an alternative to using `postinstall`.

**Reply by fbartho** [reply]
@solomonhawk I appreciate the pointer; however I choose not to use patch-package for (silly) philosophical reasons.

I generally want to discourage patches like this and adding that dep just makes such things too easy to keep around!

**Reply by asantanacu** [reply]
Same issue here and looking into the code I realized that the problem is not only for undefined values 

```
function isSerializable(refs, value, path) {
        const type = typeof value;
        if (// `null` can be serialized, but not `undefined`.
        value === null || // n.b. `bigint`, `function`, `symbol`, and `undefined` cannot be
        // serialized.
        //
        // `object` is special-cased below, as it may represent `null`, an Array,
        // a plain object, a class, et al.
        type === 'boolean' || type === 'number' || type === 'string') {
            return true;
        }
```


I can understand bigint(fails if JSON.stringify), function and symbol but undefined should be ok

**Comment by negatron99** [comment]
Using class-transformer and database models I have semantic reasons why why undefined and null are different. Null are explicitly defined as not being defined in the database. Undefined are objects that are not requested and not available in the response. 

Adding the extra steps to filter out undefined objects (another call to class-transformer) from the response makes for even more processing fluff in the controller (NestJS) just to remove something that means something.

I'm not sure why the undefined causes a problem when the basic JSON serializer just ignores undefined objects anyway?

**Comment by felipenmoura** [comment]
I simply can't believe I had to waste hours creating a `removeUndefinedsToPleaseNext` function for no reason at all :(
We get no benefits from this check.

In case someone is facing the same (very) annoying task, here is a function that will remove all `undefined` values from your props, including nested values...including arrays:

```js
function removeUndefinedsToPleaseNext (obj) {
  // cleaning an array
  if (Array.isArray(obj)) {
    const arr = obj;
    const newArr = [];
    arr.forEach((val, key) => {
      if (typeof val === 'undefined') {
        return;
      }
      if (val === Object(val)) {
        // this is an object, not a regula value
        newArr[key] = removeUndefinedsToPleaseNext(val);
      } else {
        newArr[key] = val;
      }
    });
    return newArr;
  }

  // cleaning an object
  const newObj = {};
  Object.keys(obj).forEach((key) => {
    const val = obj[key];
    if (typeof val === 'undefined') {
      return;
    }
    if (val === Object(val)) {
      // this is an object, not a regula value
      newObj[key] = removeUndefinedsToPleaseNext(val);
    } else {
      newObj[key] = val;
    }
  });
  return newObj;
}
```

Then, just apply it to your props in your getStaticProps:

```js
export async function getStaticProps({ locale, params }) {
  // ...
  return {
    props: removeUndefineds({ ... })
  };
}
```

**Comment by TheDunco** [comment]
I'd imagine the reason they didn't end up actually changing this is because Typescript is the only thing that makes it feasible to let this slide. For vanilla JS Next users, having a key simply disappear from a props object could cause serious problems.

However, if you're a TS user there is a super simple way to patch this with [patch-package](https://www.npmjs.com/package/patch-package). All you need to do is create a patch by...
1. Installing patch-package
2. Going to `node_modules/next/dist/lib/is-serializable-props.js` in your project
	3. Find the `if (type === "undefined") {...}` block in that file
	4. Replace the content of the block with `return true;`
6. Then head back to the root directory of your project and run `npx patch-package next` and you're good to go!

**Comment by Morglod** [comment]
You doing great framework for some years; and than decides to do this:
```ts
type PageProps = {
    x?: string
}

getServerSideProps = (): { props: PageProps } => ({
   x: null // <---- hello vercel, it broke types here; I need undefined, thx
})
```

Please let us use "ultaDangerouzzUndefinedServerSizeProps=true" in config and stop monkey patching

**Comment by PrinOrange** [comment]
To solve this problem, I wrote a kit package [nullify-undefined](https://www.npmjs.com/package/nullify-undefined), Convert undefined in the object to null to solve the problem of javascript object serialization in some scenarios.

```bash
npm install nullify-undefined
```
You can try it with this problem.

For example, I need to transform all undefined values into null:

```javascript
var obj = {
  a: "the value of a",
  b: {
    c: undefined,
    d: [{ e: undefined }, undefined],
  },
  f: null,
};
```

use `Nullify` function

```javascript
import Nullify from "nullify-undefined";
Nullify(obj);
```

Output:

```javascript
obj = {
  a: "the value of a",
  b: {
    c: null,
    d: [{ e: null }, null],
  },
  f: null,
};

```

This function changes the value of the object passed in, and the return value is a reference to the passed object. If you don't want to change the passed object, you can make a deep copy of the passed object. like this:

```javascript
Nullify({...obj});
```

**Reply by vsDizzy** [reply]
JSON.parse(JSON.stringify(..)) is way faster @PrinOrange
